### ISSUE TITLE
 flexible array member

### ISSUE THREAD

**Issue by ghazel** [ffi/ffi#436]
How should I declare this struct with ffi?

```c
struct blob {
    size_t length;
    unsigned char data[];
};
```

I've tried:

- `:data, [:uchar, nil]`: doesn't work (fails to make the struct)
- the old C90 trick of `:data, [:uchar, 0]` doesn't work (truncates the array to 0 no matter what)
- `:data, :pointer` doesn't work. (the value of `data` is the first `sizeof(void*)` bytes of the array, so dereferencing that obviously crashes.)

So, my guess is this is a feature request for flexible array member support.

**Comment by tduehr** [comment]
I'm not sure this is strictly possible to do from FFI. What you can do for now is declare the array of the largest size it's going to be. Just as you'd do to malloc this struct in C. Then do a read the size of the length member.

**Comment by ghazel** [comment]
Is something like this possible with FFI? https://github.com/twall/jna/issues/449#issuecomment-106891200

**Comment by tduehr** [comment]
Yes and no. If you create the struct class with the max size for the data member you can create the same blob method to do the same action. You could also easily override data and data= to automatically set or read the length member. But, the structure layout is stored in the class, iirc, so manipulating it as runtime wouldn't be all that safe.

All FFI can do past what it does, without quite a bit of heavy listing, is allow you to specify a flexible length member (probably like you were suggesting: `:data, [:uchar, nil]`. The rest will likely be up to the user.
